src/DataSource.ts

```diff
@@ -97,31 +95,65 @@ export class DataSource extends DataSourceApi<AkenzaQuery, AkenzaDataSourceConfi
             }
         }
 
-        return { data: panelData };
+        return {data: panelData};
     }
 
     async getDevices(searchString?: string): Promise<Device[]> {
-        return this.getOrganization().then((organization) => {
-            return new Promise((resolve, reject) => {
-                const params = {
-                    organizationId: organization.id,
-                    type: 'DEVICE',
-                    search: searchString,
-                };
-
-                this.executeRequest<DeviceList>(
-                    '/v3/assets',
-                    'GET',
-                    params)
-                    .subscribe({
-                        next(response: FetchResponse<DeviceList>) {
-                            resolve(response.data.content)
-                        },
-                        error(error: FetchError) {
-                            reject(DataSource.generateErrorMessage(error))
-                        }
-                    });
-            });
+        const organization = await this.getOrganization();
+        const access = await this.checkWorkspaceAccess(organization.id);
+
+        const params: AccessCheckParams = {
+            organizationId: undefined,
+            type: 'DEVICE',
+            search: searchString,
+            workspaceIds: undefined,
+        };
+
+        return new Promise((resolve, reject) => {
+            // if the api key has access to all workspaces (no restrictions) the organizationId can directly be provided
+            // to query the assets
+            if (access.all) {
+                params.organizationId = organization.id
+            } else {
+                // otherwise the filter for workspaceIds is used to restrict the asset-list to only visible devices
+                params.workspaceIds = access.ids;
+            }
+
+            this.executeRequest<DeviceList>(
+                '/v3/assets',
+                'GET',
+                params)
+                .subscribe({
+                    next(response: FetchResponse<DeviceList>) {
+                        resolve(response.data.content)
+                    },
+                    error(error: FetchError) {
+                        reject(DataSource.generateErrorMessage(error))
+                    }
+                });
+        });
+    }
+
+    async checkWorkspaceAccess(organizationId: string): Promise<AccessCheckResponse> {
```

Comment:
did you find a way to cache this?

Reply:
hm, completely forgot about that aspect...